onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,622 @@
+      }
+      require_type_change.insert(output);
+    }
+    if (has_float_outputs && all_float_outputs_have_casts && casts.size() > 0 ) {
+      VLOGS(logger, 1) << "PropagateCastsFromOutputsToInputs: Removed Cast nodes "
+                       << GatherNames<std::vector<Node*>>(casts)
+                       << " feeding the same compute node " << node->Name();
+      RemoveCastNodes(graph, casts);
+      std::unordered_set<NodeArg*> node_args;
+      for (NodeArg* input : node->MutableInputDefs()) {
+        if (input->TypeAsProto()->tensor_type().elem_type() == TensorProto::FLOAT) {
+          node_args.insert(input);
+        }
+      }
+      InsertCastNodes(graph, node_args, true);
+      ChangeTypeToFP16(require_type_change, logger);
+      VLOGS(logger, 1) << "PropagateCastsFromOutputsToInputs: Inserted Cast node to " << GatherNames(node_args);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// Expand FP16 compute regions on the graph by example float16 compute nodes,
+// propagating float32 Cast operation down the graph and propagating float16
+// Cast operations up the graph. The following functions are performed
+// 1. Fuse subgraphs
+// 2. Propagate fp32 casts forwards
+// 3. Propagate fp16 casts back
+Status PropagateCastOps::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  // First apply the transformation to the subgraphs.
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (nullptr == node_ptr)
+      continue;  // node was removed
+
+    auto& node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+  }
+
+  bool local_modified = false;
+  do {
+    if (local_modified) {
+      graph.Resolve();
+    }
+
+    local_modified = RemoveUnnecessaryCasts(graph, logger);
+
+    // Fuse subgraphs, sibling Cast nodes with same input
+    for (auto& node: graph.Nodes()) {
+      local_modified |= FuseSubgraphs(graph, &node, logger);
+    }
+
+    // Propagate FP32 Casts forward
+    for (Node& node : graph.Nodes()) {
+        local_modified |= PropagateForwards(graph, &node, logger);
+    }
+
+    local_modified |= RemoveBackToBackCasts(graph, logger);
+
+    // Propagate FP16 Casts backward
+    for (Node& node : graph.Nodes()) {
+      local_modified |= PropagateBackwards(graph, &node, logger);
+    }
+
+    // Propagate FP16 Casts from outputs to inputs
+    for (Node& node : graph.Nodes()) {
+      local_modified |= PropagateFP16CastsFromOutputsToInputs(graph, &node, logger);
+    }
+
+    // Propagate FP32 Casts from inputs to outputs
+    for (Node& node : graph.Nodes()) {
+      local_modified |= PropagateFP32CastsFromInputsToOutputs(graph, &node, logger);
+    }
+
+    modified |= local_modified;
+  } while (local_modified);
+
```

Comment:
we should also log a summary on which nodes we have placed on fp16.... especially for Level2 ops! 
This will help preventing unwanted placement.